modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java

```diff
@@ -2163,68 +2159,44 @@ public void repopulate(final ElasticsearchIndex index) throws IndexRebuildExcept
+      String currentMediapackageId;
+      String lastMediapackageId = "";
+
       do {
         try {
-          workflows = persistence.getLatestWorkflowInstances(limit, token);
-        } catch (Exception e) {
-          logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), total, current, e);
+          workflowIndexData = persistence.getWorkflowIndexData(limit, offset);
+        } catch (WorkflowDatabaseException e) {
+          logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
           throw new IndexRebuildException(index.getIndexName(), getService(), e);
         }
-        logger.debug("Got {} workflows for re-indexing", workflows.size());
-        int indexLast = workflows.size() - 1;
-        token = workflows.get(indexLast).getId();
+        offset += limit;
+        logger.debug("Got {} workflows for re-indexing", workflowIndexData.size());
 
-        for (WorkflowInstance instance : workflows) {
-          current += 1;
-          Organization organization = null;
-          try {
-            organization = organizationDirectoryService.getOrganization(instance.getOrganizationId());
-          } catch (NotFoundException e) {
-            logger.error("Found workflow with non-existing organization {}", instance.getOrganizationId());
+        for (WorkflowIndexData data : workflowIndexData) {
+          currentMediapackageId = data.getMediaPackageId();
+          if (currentMediapackageId.equals(lastMediapackageId)) {
             continue;
           }
+          else {
+            updateWorkflowInstanceInIndex(data, index);
 
-          // get metadata for index update
-          final DublinCoreCatalog episodeDublinCoreCatalog = getEpisodeDublinCoreCatalog(instance.getMediaPackage());
-
-          // get acl for active workflows.
-          // don't try this for terminated workflows since the ACLs are no longer in the working file repository and
-          // they will be overwritten later in the re-indexing process by ACLs from the asset manager anyway.
-          final AccessControlList accessControlList;
-          if (instance.getState().isTerminated()) {
-            accessControlList = new AccessControlList();
-          } else {
-            accessControlList = authorizationService.getActiveAcl(instance.getMediaPackage()).getA();
+            current += 1;
+            lastMediapackageId = currentMediapackageId;
+            logIndexRebuildProgress(logger.getSlf4jLogger(), index.getIndexName(), total, current);
           }
-
-          SecurityUtil.runAs(securityService, organization,
-                  SecurityUtil.createSystemUser(componentContext, organization), () -> {
-                    updateWorkflowInstanceInIndex(instance, accessControlList, episodeDublinCoreCatalog, index);
-                  });
-          logIndexRebuildProgress(logger.getSlf4jLogger(), index.getIndexName(), total, current);
         }
       } while (current < total);
```

Comment:
I just noticed that this is dangerous since new workflows could be added while this is running.
A better termination criteria would be something like:
```suggestion
      } while (workflowIndexData.size() <= 0);
```